src/jail_plugin.c

```diff
@@ -0,0 +1,183 @@
+ * are those of the authors and should not be interpreted as representing
+ * any official policies, either expressed or implied.
+ */
+
+#define GLIB_DISABLE_DEPRECATION_WARNINGS
+
+#include "jail_types_p.h"
+#include "jail_plugin_impl.h"
+#include "jail_plugin_p.h"
+
+#include <gutil_log.h>
+
+struct jail_plugin_priv {
+    gboolean started;
+};
+
+G_DEFINE_ABSTRACT_TYPE(JailPlugin, jail_plugin, G_TYPE_OBJECT)
+#define THIS(obj) JAIL_PLUGIN(obj)
+#define PARENT_CLASS jail_plugin_parent_class
+#define GET_CLASS(obj) G_TYPE_INSTANCE_GET_CLASS((obj), \
+    JAIL_TYPE_PLUGIN, JailPluginClass)
+
+/*==========================================================================*
+ * Internal interface
+ *==========================================================================*/
+
+JailPlugin*
+jail_plugin_ref(
+    JailPlugin* self)
+{
+    if (G_LIKELY(self)) {
+        g_object_ref(THIS(self));
+    }
+    return self;
+}
+
+void
+jail_plugin_unref(
+    JailPlugin* self)
+{
+    if (G_LIKELY(self)) {
+        g_object_unref(THIS(self));
+    }
+}
+
+gboolean
+jail_plugin_start(
+    JailPlugin* self,
+    SailJail* jail)
+{
+    if (G_LIKELY(self)) {
+        JailPluginPriv* priv = self->priv;
+
+        /* Don't start twice */
+        if (priv->started) {
+            /* Already started */
+            return TRUE;
+        } else {
+            GError* error = NULL;
+
+            if (GET_CLASS(self)->start(self, jail, &error)) {
+                /* Started successfully */
+                priv->started = TRUE;
+                return TRUE;
+            }
+            GERR("Plugin \"%s\" failed to start: %s", self->desc->name,
+                GERRMSG(error));
+            g_error_free(error);
+        }
+    }
+    return FALSE;
+}
+
+void
+jail_plugin_stop(
+    JailPlugin* self)
+{
+    if (G_LIKELY(self)) {
+        JailPluginPriv* priv = self->priv;
+
+        /* Only stop if started */
+        if (priv->started) {
+            GET_CLASS(self)->stop(self);
+            priv->started = FALSE;
+        }
+    }
+}
+
+/*==========================================================================*
+ * Internals
+ *==========================================================================*/
+
+static
+gboolean
+jail_plugin_default_start(
+    JailPlugin* self,
+    SailJail* jail,
+    GError** error)
+{
+    return FALSE;
+}
+
+static
+void
+jail_plugin_default_stop(
+    JailPlugin* self)
+{
+}
+
+static
+void
+jail_plugin_init(
+    JailPlugin* self)
+{
+    self->priv = G_TYPE_INSTANCE_GET_PRIVATE(self, JAIL_TYPE_PLUGIN,
+        JailPluginPriv);
+}
+
+static
+void
+jail_plugin_dispose(
+    GObject* object)
+{
+    JailPlugin* self = THIS(object);
+
+    jail_plugin_stop(self);
+    G_OBJECT_CLASS(PARENT_CLASS)->dispose(object);
+}
+
+static
+void
+jail_plugin_class_init(
+    JailPluginClass* klass)
+{
+    GObjectClass* object_class = G_OBJECT_CLASS(klass);
+
+    g_type_class_add_private(klass, sizeof(JailPluginPriv));
```

Comment:
Since already requiring Glib 2.44, which already has the G_ADD_PRIVATE (and G_DEFINE_ABSTRACT_TYPE_WITH_PRIVATE), I'd suggest already changing away from this deprecated private handling; granted, that would pretty much force renaming the private type to use full Private suffix.

Reply:
Agree, will do!

Reply:
Switched to `G_ADD_PRIVATE`